certformat.go

```diff
@@ -0,0 +1,198 @@
+package certinfo
+
+import (
+	"bytes"
+	"crypto/dsa"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"encoding/asn1"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/smallstep/cli/pkg/x509"
+	"golang.org/x/crypto/ed25519"
+)
+
+// formatBuffer is a helper to write using sprintf.
+type formatBuffer struct {
+	bytes.Buffer
+}
+
+// Writef writes a string formated using fmt.Sprintf.
+func (b *formatBuffer) Writef(format string, args ...interface{}) (int, error) {
+	return b.Buffer.WriteString(fmt.Sprintf(format, args...))
+}
+
+type certificateShort struct {
+	Type               string
+	PublicKeyAlgorithm string
+	SerialNumber       string
+	Subject            string
+	Issuer             string
+	SANs               []string
+	Provisioner        *provisioner
+	NotBefore          time.Time
+	NotAfter           time.Time
+}
+
+type provisioner struct {
+	ID   string
+	Name string
+}
+
+func newCertificateShort(cert *x509.Certificate) *certificateShort {
+	var typ string
+	if cert.IsCA {
+		if cert.CheckSignatureFrom(cert) == nil {
+			typ = "Root CA"
+		} else {
+			typ = "Intermediate CA"
+		}
+	} else {
+		typ = "TLS"
+	}
+
+	return &certificateShort{
+		Type:               typ,
+		PublicKeyAlgorithm: getPublicKeyAlgorithm(cert.PublicKeyAlgorithm, cert.PublicKey),
+		SerialNumber:       abbreviated(cert.SerialNumber.String()),
+		Subject:            cert.Subject.CommonName,
+		Issuer:             cert.Issuer.CommonName,
+		SANs:               getSANs(cert),
+		Provisioner:        getProvisioner(cert),
+		NotBefore:          cert.NotBefore,
+		NotAfter:           cert.NotAfter,
+	}
+}
+
+// String returns the certificateShort formated as a string.
+func (c *certificateShort) String() string {
+	var buf formatBuffer
+	buf.Writef("X.509v3 %s Certificate (%s) [Serial: %s]\n", c.Type, c.PublicKeyAlgorithm, c.SerialNumber)
+	buf.Writef("  Subject:     %s\n", c.Subject)
+	for _, s := range c.SANs {
+		buf.Writef("               %s\n", s)
+	}
+	buf.Writef("  Issuer:      %s\n", c.Issuer)
+	if c.Provisioner != nil {
+		buf.Writef("  Provisioner: %s [ID: %s]\n", c.Provisioner.Name, c.Provisioner.ID)
+	}
+	buf.Writef("  Valid from:  %s\n", c.NotBefore.Format(time.RFC3339))
+	buf.Writef("          to:  %s\n", c.NotAfter.Format(time.RFC3339))
+	return buf.String()
+}
+
+type certificateRequestShort struct {
+	PublicKeyAlgorithm string
+	Subject            string
+	SANs               []string
+}
+
+func newCertificateRequestShort(cr *x509.CertificateRequest) *certificateRequestShort {
+	var sans []string
+	for _, s := range cr.DNSNames {
```

Comment:
How come you're not using `getSans()` here?

Reply:
because they are different types